### PR TITLE
Change precision on all selected nodes

### PR DESCRIPTION
**Pull request by samhardin**
### Purpose of this PR
Allows users to change precision on all selected nodes simultaneously.
Precision_On_Selected

Properly populating downstream:
Precision_Toggle2

Also, placed the 'Convert To' options in a submenu:
New_Contextual

---
### Release Notes
Please add any useful notes about the feature/fix that might be helpful for others.

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: What did you do?
- [x] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=sg%2Ftoggle-selected-precision&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Ftoggle-selected-precision

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High? Low

**Halo Effect**: None, Low, Medium, High? Low

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 1]
```diff
@@ -345,6 +372,30 @@ public void SetPreviewExpandedOnSelection(bool state)
             }
         }
 
+        public void SetNodePrecisionOnSelection(Precision inPrecision)
+        {
+            var editorView = GetFirstAncestorOfType<GraphEditorView>();
+            IEnumerable<MaterialNodeView> nodes = selection.Where(x => x is MaterialNodeView).Select(x => x as MaterialNodeView);
+
+            graph.owner.RegisterCompleteObjectUndo("Set Precisions");
+            var nodeList = this.Query<MaterialNodeView>().ToList();
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: This line seems unnecessary.

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 2]
```diff
@@ -191,6 +192,32 @@ public override void BuildContextualMenu(ContextualMenuPopulateEvent evt)
...
+                if (selectedNode.node.precision != Precision.Half)
+                    halfPrecisionAction = DropdownMenuAction.Status.Normal;
+            }
+
+            // Create the menu options
+            evt.menu.AppendAction("Precision/Inherit", _ => SetNodePrecisionOnSelection(Precision.Inherit), (a) => inheritPrecisionAction);
+            evt.menu.AppendAction("Precision/Float", _ => SetNodePrecisionOnSelection(Precision.Float), (a) => floatPrecisionAction);
+            evt.menu.AppendAction("Precision/Half", _ => SetNodePrecisionOnSelection(Precision.Half), (a) => halfPrecisionAction);
+
+            //evt.menu.AppendSeparator();
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: Don't leave commented code in commits.

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs (hunk 1 above)
Comment (samhardin) [reply com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: You are completely right!

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs (hunk 2 above)
Comment (samhardin) [reply com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: Oh! Good catch! I missed that on my cleaning pass.

**Comment by samhardin** [comment]
> The only thing I would add is consider fixing the menu separators in this PR?
> (https://unity.slack.com/archives/G7HQCPV71/p1564443859027900)

I can group the sub menu options, but sadly all ones above that come from GraphView in trunk.

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 3]
```diff
@@ -345,6 +369,29 @@ public void SetPreviewExpandedOnSelection(bool state)
             }
         }
 
+        public void SetNodePrecisionOnSelection(Precision inPrecision)
+        {
+            var editorView = GetFirstAncestorOfType<GraphEditorView>();
+            IEnumerable<MaterialNodeView> nodes = selection.Where(x => x is MaterialNodeView).Select(x => x as MaterialNodeView);
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: `selection.Where(x => x is MaterialNodeView mnv && mnv.node.canSetPrecision)` and then we don't need the canSetPrecision check in the for loop.

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 4]
```diff
@@ -345,6 +369,29 @@ public void SetPreviewExpandedOnSelection(bool state)
...
+
+            graph.owner.RegisterCompleteObjectUndo("Set Precisions");
+            editorView.colorManager.SetNodesDirty(nodes);
+
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs]: Early out here if nodes is empty

**Comment by Kink3d** [comment]
Rebuilt at #4536

**Pull request closed without merging**